### DISCUSSION TITLE
Does README.md have max characters?

### DISCUSSION THREAD

**Discussion by halfrost** [community/community#23920] [New to GitHub]
<p>In my repo (<a href="https://github.com/halfrost/LeetCode-Go" rel="nofollow noopener">https://github.com/halfrost/LeetCode-Go</a>) README doesn’t render complete.</p>
<p>The last line is incomplete:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/1/185c94762ea12ac1f9a1a4337edef86202c3c77e.png" data-download-href="/uploads/short-url/3tvOgZB998KChnHBgO3jqcNow90.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/10825609/181080020-8a2339a5-abd5-463f-b5ec-d0ef394ec357.png" alt="image" data-base62-sha1="3tvOgZB998KChnHBgO3jqcNow90" width="690" height="108" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/1/185c94762ea12ac1f9a1a4337edef86202c3c77e_2_690x108.png, https://aws1.discourse-cdn.com/github/original/2X/1/185c94762ea12ac1f9a1a4337edef86202c3c77e.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/1/185c94762ea12ac1f9a1a4337edef86202c3c77e.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/1/185c94762ea12ac1f9a1a4337edef86202c3c77e_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">872×137 10.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>You can see this problem in this commit: <a href="https://github.com/halfrost/LeetCode-Go/tree/7989003450a23bf1c27a34201f124cb2745d7096" rel="nofollow noopener">https://github.com/halfrost/LeetCode-Go/tree/7989003450a23bf1c27a34201f124cb2745d7096</a>, the last line of README.md.</p>
<p>In my local, it looks correct.</p>
<pre><code class="lang-auto">
| Title | Solution | Difficulty | Time | Space | 收藏 |
| ----- | :--------: | :----------: | :----: | :-----: |:-----: |
|[973. K Closest Points to Origin](https://leetcode.com/problems/k-closest-points-to-origin/)| [Go](https://github.com/halfrost/LeetCode-Go/tree/master/leetcode/0973.K-Closest-Points-to-Origin)| Medium | O(n log n)| O(log n) ||
|[976. Largest Perimeter Triangle](https://leetcode.com/problems/largest-perimeter-triangle/)| [Go](https://github.com/halfrost/LeetCode-Go/tree/master/leetcode/0976.Largest-Perimeter-Triangle)| Easy | O(n log n)| O(log n) ||
|[1030. Matrix Cells in Distance Order](https://leetcode.com/problems/matrix-cells-in-distance-order/)| [Go](https://github.com/halfrost/LeetCode-Go/tree/master/leetcode/1030.Matrix-Cells-in-Distance-Order)| Easy | O(n^2)| O(1) ||
|[1054. Distant Barcodes](https://leetcode.com/problems/distant-barcodes/)| [Go](https://github.com/halfrost/LeetCode-Go/tree/master/leetcode/1054.Distant-Barcodes)| Medium | O(n log n)| O(log n) |❤️|

</code></pre>
<p>This file has no grammatical errors, but GitHub can not render it complete.</p>
<p>Does this README.md trigger max characters? Why GitHub renders it incomplete? It’s a bug or not?</p>

**Comment by halfrost** [comment]
<p>In my local, it looks correct:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/1/1bb6b76d81795a032f85b119673701b188081ae0.png" data-download-href="/uploads/short-url/3XamjX58P3opyGsY4oBiCRyf1GE.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/10825609/181080022-7d4066d7-c7d6-4994-a8b7-03489126776e.png" alt="image" data-base62-sha1="3XamjX58P3opyGsY4oBiCRyf1GE" width="690" height="191" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/1/1bb6b76d81795a032f85b119673701b188081ae0_2_690x191.png, https://aws1.discourse-cdn.com/github/optimized/2X/1/1bb6b76d81795a032f85b119673701b188081ae0_2_1035x286.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/1/1bb6b76d81795a032f85b119673701b188081ae0_2_1380x382.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/1/1bb6b76d81795a032f85b119673701b188081ae0_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1566×434 51.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by nethgato** [comment, marked as answer]
<p>Hey <a class="mention" href="/halfrost">@halfrost</a> o/</p>
<p>That is strange, and I wish I could say I have a substantial answer for you straight away, but I don’t =(</p>
<p>I’ve scoured each our docs, internal ticket history, slack history, and even our own codebase. I’ve been unable to find any documented (or coded) character limits on markdown files; <code>README</code> or otherwise.</p>
<p>I have seen some internal comments that suggest, no; we do not have a character limit. But because of the experience you see, there very well may be a limit. Since I’ve spent some time on this, I’m hesitant to keep digging before hearing from you.</p>
<p>Since that README is so extensive, should we discuss the usability of the page? Is there functionality elsewhere that makes navigating such a large page easy for your users?</p>
<p>While I can definitely go back here and ping some folks, I would wonder why we’re not breaking these segments into their own files, or even into gists, to store that information. Rather than having it all live in the README.</p>
<p>I can assume that just having that bit of knowledge as an official stance is valuable, but perhaps it may be best in your case to consider organizing the information in another way.</p>
<p>Do please let me know what you think, and if any other users have run into this, curious to hear what people think and if I need to go hunt down that person potentially responsible for this unforeseen limit/bug.</p>
<p>Cheers :bow:</p>

**Comment by halfrost** [comment]
<p>Hi, <a class="mention" href="/nethgato">@nethgato</a></p>
<p>Thanks for your reply. I try some solutions which contain different words in the README. I am sure this behavior, trimming some characters, is made by Github render.</p>
<p>However, I support your opinions. If README is too long, this is really unfriendly to the readers. I have split it into different parts. In the latest commit, everything is fine.</p>
<p>Finally, thank you for patiently answering this strange question of mine. Maybe I waste your time. :sweat_smile:</p>

**Comment by nethgato** [comment]
<p>Hey <a class="mention" href="/halfrost">@halfrost</a> – absolutely <em>not</em> a waste of time! Thank you for replying back and being willing to rethink your approach to the layout of your README.</p>
<p>It really would be great to have an official answer, but it was rather difficult to track down. I had hoped to avoid pinging too many folks in an attempt to find that answer. If there’s any more interest in finding it, I can re-engage with folks that may know better than myself. Surprisingly, it was more difficult than I thought it would be =D</p>

**Comment by lukasmasuch** [comment]
<p>Hey <a class="mention" href="/nethgato">@nethgato</a> ,</p>
<p>I also run into some kind of README limitation as well. But it only seems to apply for the Readme displayed on the main repo page, and not if you open the same file directly. The readme rendering here is cut off: <a href="https://github.com/ml-tooling/best-of-ml-python/tree/2020.12.27" class="inline-onebox" rel="noopener nofollow ugc">GitHub - ml-tooling/best-of-ml-python at 2020.12.27</a> while the rendering here for the same file is fine: <a href="https://github.com/ml-tooling/best-of-ml-python/blob/2020.12.27/README.md" class="inline-onebox" rel="noopener nofollow ugc">best-of-ml-python/README.md at 2020.12.27 · ml-tooling/best-of-ml-python · GitHub</a> . I did a few experiments and figured out that the limitation seems to be around 500k characters. However, this might be just a coincidence and maybe there is just a rendering timeout for the main repo page.</p>
<p>Update: I think I figured out the limit. The readme on the main page only renders the first 512 KB of the README.md file. This is around 500k characters (+/- a few thousand).</p>

**Comment by lukasmasuch** [comment]
<p>Here is another example with just using lorem ipsum text: The text on the main repo page is cut off<br>
<a href="https://github.com/LukasMasuch/readme-limit-test" class="inline-onebox" rel="noopener nofollow ugc">GitHub - LukasMasuch/readme-limit-test: Testing Github readme rendering limitations</a>  while the rendering for the file is fine: <a href="https://github.com/LukasMasuch/readme-limit-test/blob/main/README.md" class="inline-onebox" rel="noopener nofollow ugc">readme-limit-test/README.md at main · LukasMasuch/readme-limit-test · GitHub</a></p>

**Comment by kaiwensun** [comment]
<p>I experienced the same problem. <a href="https://github.com/kaiwensun/leetcode/tree/cd1f682e8e8bd17c1bb67ee0c8c260b4d4537d19" class="inline-onebox" rel="noopener nofollow ugc">GitHub - kaiwensun/leetcode at cd1f682e8e8bd17c1bb67ee0c8c260b4d4537d19</a> is the last healthy version of README. After I insert some more text in the middle of the readme, the tail of the readme is truncated as in <a href="https://github.com/kaiwensun/leetcode/tree/6922b8e0521641f5b570fe83a77e2c5243473788" class="inline-onebox" rel="noopener nofollow ugc">GitHub - kaiwensun/leetcode at 6922b8e0521641f5b570fe83a77e2c5243473788</a> .</p>
<p>Same as what <a class="mention" href="/lukasmasuch">@LukasMasuch</a> mentioned,</p>
<blockquote>
<p>it only seems to apply for the Readme displayed on the main repo page, and not if you open the same file directly</p>
</blockquote>

**Comment by kaiwensun** [comment]
<p>We’ve seen the complaints of this same issue from multiple users.</p>
<p>I am new to GitHub community. What is the correct way to get attention from GitHub staff to get this fixed? If here is not the correct channel, can you point me to the correct one?</p>
<p>Thank you.</p>

**Reply by Ebazhanov** [reply]
Hello, 

the same happens here as well https://github.com/Ebazhanov/linkedin-skill-assessments-quizzes

**Comment by kaiwensun** [comment]
Hi all,

I opened a [private support ticket](https://support.github.com/ticket/personal/0/1295432) to GitHub support team in Sept 2021. The short answer is **"GitHub currently limit blob displaying to around 500 KB of data"** when the readme is displayed at the home page of the repo.

The following is the full reply from the GitHub support team.

> Hey Kaiwen,
>
> I reached out to our engineering team to figure out what is going on and we have an explanation for the issue you're facing.
>
> We currently limit blob displaying to around 500 KB of data, anything more we truncate in the UI. You can view the raw by clicking through.
>
> Your README.md file is 513 KB and hence would be truncated on the UI. However, as you may have noticed, accessing the blob directly works just fine i.e https://github.com/kaiwensun/leetcode/blob/master/README.md
>
> I'm sorry this answer may not provide a resolution to your query, but please understand such system limits are put in place to enhance fast & seamless user experience on the UI.
> 
> Please feel free to reach out to me if you have further questions about this request.
>
> All the best,
> Malcolm

**Comment by kaiwensun** [comment]
Looks like GitHub recently reduced the max limit even more! WITHOUT NOTICE! This is a BREAKING CHANGE!

Also, the workaround "However, as you may have noticed, accessing the blob directly works just fine" does not work anymore.

Cutting a [new private support ticket](https://support.github.com/ticket/personal/0/2156416) to GitHub.
